src/odm-collection.hpp

```diff
@@ -0,0 +1,727 @@
+template <class T>
+class out_archiver_mock {
+   public:
+    out_archiver_mock(std::ostream& os) : _os(&os) {
+    }
+
+    void operator()(T obj) {
+        bsoncxx::builder::basic::document basic_builder{};
+        basic_builder.append(bsoncxx::builder::basic::kvp("a", obj.a));
+        basic_builder.append(bsoncxx::builder::basic::kvp("b", obj.b));
+        basic_builder.append(bsoncxx::builder::basic::kvp("c", obj.c));
+        auto v = basic_builder.view();
+        _os->write(reinterpret_cast<const char*>(v.data()), v.length());
+    }
+
+   private:
+    std::ostream* _os;
+};
+
+// A mock class for testing de-serialization w/o an actual BSON serializer class.
+// This class is templated, but requiers the given type to ahve primitize fields a, b, c.
+// This will be removed once cereal::BSONArchiver is merged in
+template <class T>
+class in_archiver_mock {
+   public:
+    in_archiver_mock(std::istream& is) : _is(&is) {
+    }
+
+    void operator()(T& obj) {
+        // get document size
+        int size = 0;
+        _is->read(reinterpret_cast<char*>(&size), 4);
+        uint8_t* data = new uint8_t[size];
+        // read BSON data into document view
+        _is->seekg(0);
+        _is->read(reinterpret_cast<char*>(data), size);
+        auto doc_view = bsoncxx::document::view(data, size);
+        // Fill object with fields
+        obj.a = doc_view["a"].get_int32();
+        obj.b = doc_view["b"].get_int32();
+        obj.c = doc_view["c"].get_int32();
+    }
+
+   private:
+    std::istream* _is;
+};
+
+/**
+ * Converts a serializable object into a BSON document value
+ * TODO This is not very clean and kind of inefficient. Maybe we should be passing references
+ * into
+ * the bson_ostream callback?
+ * @tparam T   A type that is serializable to BSON using a BSONArchiver.
+ * @param  obj A serializable object
+ * @return     A BSON document value representing the given object.
+ */
+template <class T>
+bsoncxx::document::value to_document(const T& obj) {
+    bsoncxx::document::value doc(nullptr, 0, [](uint8_t*) {});
+    bson_mapper::bson_ostream bos([&doc](bsoncxx::document::value v) { doc = std::move(v); });
+    // TODO cereal::BSONOutputArchiver archiver(bos);
+    out_archiver_mock<T> archive(bos);
+    archive(obj);
+    return doc;
+}
+
+/**
+* Converts a bsoncxx document view to an object of the templated type through deserialization.
+*
+* @tparam T A type that is serializable using a BSONArchiver
+* @param v A BSON document view. If the BSON document does not match the schema of type T,
+* BSONArchiver will throw a corresponding exception.
+* TODO what does BSONArchiver throw?
+* @return Returns by value an object that corresponds to the given document view.
+*/
+template <class T>
+T to_obj(bsoncxx::document::view v) {
+    bson_mapper::bson_istream bis(v);
+    // TODO cereal::BSONInputArchive archive(bis);
+    in_archiver_mock<T> archive(bis);
+    T obj;
```

Comment:
This requires T is default constructible. If we want to enforce this (I bet we do...), we should static assert. See [view_or_value](https://github.com/mongodb/mongo-cxx-driver/blob/master/src/bsoncxx/view_or_value.hpp#L41-L45) for an example of this.